### PR TITLE
Fix partition dtypes so that `to_pandas` always has correct dtypes

### LINKED ISSUE #1165: Incorrect dtype when read_csv reads a file big enough
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:

Ubuntu 19.10

- **Modin version** (`modin.__version__`):

0.7.2+37.g52a5b16 on Ray 0.8.3

- **Python version**:

Python 3.7.5

- **Code we can use to reproduce**:

```python
if __name__ == "__main__":
    #import pandas as pd
    import modin.pandas as pd
    from collections import OrderedDict
    cols = [
        "loan_id",
        "monthly_reporting_period",
        "servicer",
        "interest_rate",
        "current_actual_upb",
        "loan_age",
        "remaining_months_to_legal_maturity",
        "adj_remaining_months_to_maturity",
        "maturity_date",
        "msa",
        "current_loan_delinquency_status",
        "mod_flag",
        "zero_balance_code",
        "zero_balance_effective_date",
        "last_paid_installment_date",
        "foreclosed_after",
        "disposition_date",
        "foreclosure_costs",
        "prop_preservation_and_repair_costs",
        "asset_recovery_costs",
        "misc_holding_expenses",
        "holding_taxes",
        "net_sale_proceeds",
        "credit_enhancement_proceeds",
        "repurchase_make_whole_proceeds",
        "other_foreclosure_proceeds",
        "non_interest_bearing_upb",
        "principal_forgiveness_upb",
        "repurchase_make_whole_proceeds_flag",
        "foreclosure_principal_write_off_amount",
        "servicing_activity_indicator",
    ]

    dtypes = OrderedDict(
        [
            ("loan_id", "int64"),
            ("monthly_reporting_period", "datetime64"),
            ("servicer", "category"),
            ("interest_rate", "float64"),
            ("current_actual_upb", "float64"),
            ("loan_age", "float64"),
            ("remaining_months_to_legal_maturity", "float64"),
            ("adj_remaining_months_to_maturity", "float64"),
            ("maturity_date", "datetime64"),
            ("msa", "float64"),
            ("current_loan_delinquency_status", "int32"),
            ("mod_flag", "category"),
            ("zero_balance_code", "category"),
            ("zero_balance_effective_date", "datetime64"),
            ("last_paid_installment_date", "datetime64"),
            ("foreclosed_after", "datetime64"),
            ("disposition_date", "datetime64"),
            ("foreclosure_costs", "float64"),
            ("prop_preservation_and_repair_costs", "float64"),
            ("asset_recovery_costs", "float64"),
            ("misc_holding_expenses", "float64"),
            ("holding_taxes", "float64"),
            ("net_sale_proceeds", "float64"),
            ("credit_enhancement_proceeds", "float64"),
            ("repurchase_make_whole_proceeds", "float64"),
            ("other_foreclosure_proceeds", "float64"),
            ("non_interest_bearing_upb", "float64"),
            ("principal_forgiveness_upb", "float64"),
            ("repurchase_make_whole_proceeds_flag", "category"),
            ("foreclosure_principal_write_off_amount", "float64"),
            ("servicing_activity_indicator", "category"),
        ]
    )
    all_but_dates = {
        col: valtype for (col, valtype) in dtypes.items() if valtype != "datetime64"
    }
    dates_only = [
        col for (col, valtype) in dtypes.items() if valtype == "datetime64"
    ]
    df = pd.read_csv("perf.txt",
                     delimiter="|", names=cols, dtype=all_but_dates, parse_dates=dates_only)
    print(df["servicer"])
```

### Describe the problem

The bug depends on dataset size and doesn't happen when file is short. I couldn't reproduce it on 10 records but it is reproducible on 100 records for me. It is also not reproducible on MODIN_ENGIN=python, happens only Ray or Dask.

The problem is with the type of column. It should be `category` but on Ray or Dask it appears as `object` which doesn't allow to get attribute `cat`.

This is dataset that works for me. [perf.txt](https://github.com/modin-project/modin/files/4390427/perf.txt)
Original data can be downloaded from this site https://rapidsai.github.io/demos/datasets/mortgage-data

### PR DESCRIPTION
**Pull request by devin-petersohn**
* Resolves #1165
* Add a `read` function to base class of all readers `FileReader`
* Change implementation of `read` to `_read` for the children classes
* Apply Categorical `dtype` to all partitions after the dataframe has
  been read

Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1165 
- [ ] tests added and passing

### CODE REVIEW

File: modin/data_management/factories.py [hunk 1]
```diff
@@ -18,6 +18,10 @@
 from modin import __execution_engine__ as execution_engine
 from modin import __partition_format__ as partition_format
 
+import pandas
```
Comment (devin-petersohn) [inline modin/data_management/factories.py:21]: I don't think these are getting used at the moment. We should remove them if they are not used.

File: modin/engines/base/io/file_reader.py [hunk 2]
```diff
@@ -23,6 +24,32 @@ class FileReader:
     frame_partition_cls = None
     query_compiler_cls = None
 
+    @classmethod
+    def read(cls, *args, **kwargs):
+        query_compiler = cls._read(*args, **kwargs)
+        # TODO (devin-petersohn): Make this section more general for non-pandas kernel
```
Comment (devin-petersohn) [inline modin/engines/base/io/file_reader.py:30]: Probably we need to open an issue for this one. At least omnisci will need this to be changed probably, but we can fix in a later PR.

File: modin/data_management/factories.py (hunk 1 above)
Comment (gshimansky) [reply modin/data_management/factories.py:21]: Pandas name is used on the next line to define `pandas.CategoricalDtype`.

File: modin/engines/base/io/file_reader.py (hunk 2 above)
Comment (gshimansky) [reply modin/engines/base/io/file_reader.py:30]: Ok #1504

**Pull request merged**